dyc/top.py

```diff
@@ -0,0 +1,140 @@
+import sys
+import re
+import fileinput
+import copy
+import linecache
+import click
+from .utils import (
+    get_leading_whitespace,
+    BlankFormatter,
+    get_indent,
+    add_start_end,
+    get_file_lines,
+)
+from .base import Builder
+
+
+class TopBuilder(Builder):
+    def __init__(self, filename, config, placeholders=False):
+        self.filename = filename
+        self.config = config
+        self.placeholders = placeholders
+
+    def initialize(self):
+        """
+        This method overrides the method in the builder base.
+        """
+        pass
+
+    def validate(self):
+        """
+        An abstract validator method that checks if the top is
+        still valid and gives the final decision
+        """
+        if not self.is_top_file_documented():
+            confirmed = click.confirm(
+                "Do you want to document top of file {}?".format(
+                    click.style(self.filename, fg="green")
+                )
+            )
+            if confirmed:
+                return True
+
+        return False
+
+    def is_top_file_documented(self):
+        """
+        A boolean function that determines weather the first line has
+        a docstring or not
+        """
+        returned = False
+        line = linecache.getline(self.filename, 1)
+        if self.config.get("open") in line:
+            returned = True
+        linecache.clearcache()
+        return returned
+
+    def prompts(self):
+        self._prompt_docstring()
+        # TODO: Prompt args here as well
+
+    def _prompt_docstring(self):
+
+        self.Top_docstring = ""
+        if self.validate():
+            if self.placeholders:
+                self.Top_docstring = "<docstring>"
+            else:
+                name = click.style(self.filename, fg="green")
+                self.Top_docstring = click.prompt("\n({}) Top docstring ".format(name))
+
+    def apply(self):
+        if self.validate():
+            self.polish()
+            final_result = self.wrap_doc_strings(self.result)
+
+            for i, line in enumerate(fileinput.input(self.filename, inplace=True)):
+                if i == 0:
+                    sys.stdout.write(final_result + "\n" + line)
+                else:
+                    sys.stdout.write(line)
+
+    def wrap_strings(self, words):
+        """
+        Compact how many words should be in a line
+        Parameters
+        ----------
+        str words: docstring given
+        """
+        subs = []
+        n = self.config.get("words_per_line")
+        for i in range(0, len(words), n):
+            subs.append(" ".join(words[i : i + n]))
+        return "\n".join(subs)
+
+    def confirm(self, polished):
```

Comment:
That looks great. I'm just seeing a possibility of refactoring with the Methods process. It's a non blocker now. and I'd prefer to have it in a separate PR after delivering the business rule.